packages/snapshot-listener/src/utils/methodNotAllowed.ts

```diff
@@ -0,0 +1,36 @@
+import {
+  Express,
+  NextFunction,
+  Request,
+  RequestHandler,
+  Response,
+  Router,
+} from 'express';
+import { HttpMethod } from 'common-common/src/types';
+
+const routesMethods: { [key: string]: string[] } = {};
+
+export const registerRoute = (
+  router: Express,
+  method: HttpMethod,
+  path: string,
+  ...handlers: RequestHandler[]
+) => {
+  router[method](path, ...handlers);
+  if (!routesMethods[path]) routesMethods[path] = [];
+  routesMethods[path].push(method.toUpperCase());
+};
+
+export const methodNotAllowedMiddleware = () => {
+  return (req: Request, res: Response, next: NextFunction) => {
+    const methods = routesMethods[req.path];
+    if (methods && !methods.includes(req.method)) {
+      res.set('Allow', methods ? methods.join(', ') : '');
+      return res
+        .status(405)
+        .json({ result: 'Method Not Allowed', status: 405 });
+    } else {
+      return next();
+    }
+  };
```

Comment:
Since this code has a large surface area of usage and touches on critical parts of the API, I'd recommend:
1.) Documentation is added here (ideally TSDoc format) to make it clear what the purpose of this code is (it's not immediately clear without the PR context IMO)
2.) A unit test is added to ensure that edgecases are handled and future regressions are less likely

Reply:
There's some difficulty with creating good unit tests for this. For example, I can't really create a unit test that tests for arbitrary routes that DON'T use the `registerRoute` function. Unless I am mistaken, the best I can do is create tests for some (or all though this seems excessive) routes to see if they return 405 errors for incorrect methods.

Reply:
Understood.

Not looking for perfection here by any means, just some barebones coverage. Would it be possible to use mocks to add a sanity check for the cases of "is allowed" and "not allowed"– just for a single route?

If not, no worries.

Reply:
Not sure about mocks. I'll look into mocking otherwise I'll just add a test case in the API integration tests. The case of "is allowed" are all already tested i.e. regular route testing. The case for "not allowed" is what I can add a test for i.e. make a request to a route with incorrect method and expect 405 with `ALLOWED` header properly set.

Reply:
@timolegros status on outcomes from this conversation (tests / docs)? we want to get this PR in, and this is the last outstanding Q.

Reply:
Test and docs added but now blocked by #4449. Converting to draft until unblocked and conflicts resolved.

Reply:
Unblocked since #4449 was merged. @rbennettcw please resolve this if you have no further comments.

Reply:
Normally would look to @rbennettcw to resolve his own comment here, however he is OOO today so will go ahead and resolve to get this in as mentioned yesterday.